### ISSUE TITLE
AnnotationFileLoader fail with anonymous class inside a trait

### ISSUE THREAD

**Issue by rmdefi** [symfony/symfony#25796]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4.3

The class Symfony\Component\Routing\Loader\AnnotationFileLoader does not properly parse a file containing a trait which define an anonymous class.

Here is a minimal example that trigger the bug:

```php
    namespace AppBundle\Controller;
    trait TestTrait {
      protected function createClass() {
        return new class {
          //We need one T_STRING token after the class keyword to trigger the bug
          public function f() {}
        };
      }
    }
```

If I put that in a the file src/AppBundle/Controller/TestTrait.php of my symfony project (which is scanned for annotation by the router component), I get a ReflectionException:

    Class AppBundle\Controller\f does not exist
    Exception trace:
     ReflectionClass->__construct() at /home/renaud/sites/Bruxelles_formation/vendor/symfony/symfony/src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php:62
     Symfony\Component\Routing\Loader\AnnotationDirectoryLoader->load() at /home/renaud/sites/Bruxelles_formation/vendor/symfony/symfony/src/Symfony/Component/Config/Loader/FileLoader.php:153
     Symfony\Component\Config\Loader\FileLoader->doImport() at /home/renaud/sites/Bruxelles_formation/vendor/symfony/symfony/src/Symfony/Component/Config/Loader/FileLoader.php:90
     Symfony\Component\Config\Loader\FileLoader->import() at /home/renaud/sites/Bruxelles_formation/vendor/symfony/symfony/src/Symfony/Component/Routing/Loader/YamlFileLoader.php:161
    [...]

Looking at the source of the `findClass()` method (which gives the class name in `AnnotationDirectoryLoader` at line 61), it looks for the first class keyword in the file.
This is generally fine for a file that actually define a class, because even if we have anonymous classes inside the file, it would stop at the actual class definition.
But for a trait it continue looking until it eventually find a `class` keyword, and if it does the first T_STRING token found after is supposed to be the class name.

I am not sure what a proper fix could be.
The ideal fix would be to detect that the `class` keyword is actually used in an anonymous class context and then ignore it. It is the case if there is a `new` keyword before `class`, but there may be some valid tokens between them (at least comments), so I don't know if this can be done in a simple and reliable way.

A quick fix would be to detect the `trait` keyword and stop parsing the file if it is found, not sure if it is acceptable. And it would miss other case where an anonymous class is used outside of an actual class (like in a file that only define functions).

For the time being, my workaround is to put the traits outside of the directories scanned by the router.

**Comment by sroze** [comment]
I can't reproduce the issue you describe but do reproduce the following (expected) exception:

> The autoloader expected class "App\Controller\TestTrait" to be defined in file "/Users/sroze/my-symfony/vendor/composer/../../src/Controller/TestTrait.php". The file was found but the class was not in it, the class name or namespace probably has a typo in /Users/sroze/my-symfony/config/services.yaml (which is loaded in resource "/Users/sroze/my-symfony/config/services.yaml").

**Comment by rmdefi** [comment]
Which version of symfony did you try it with?

I don't understand how you end up with that result, but this is still wrong, because TestTrait is a trait, not a class, so `findClass()` should not return it.

**Comment by rmdefi** [comment]
Also, I am testing this with PHP 7.0.27

**Comment by pierredup** [comment]
I can reproduce this issue and created a test project at https://github.com/pierredup/sf-25796 (Exception message is `Class AppBundle\Controller\f does not exist in /Users/pierre/projects/test/router-test/src/AppBundle/Controller/`)

**Comment by pierredup** [comment]
This has been fixed in #25801
